neurolib/models/multimodel/builder/base/network.py

```diff
@@ -0,0 +1,717 @@
+
+    @staticmethod
+    def _prepare_mass_parameters(param, num_nodes, native_type=dict):
+        """
+        Prepare mass parameters for the network. I.e. extend to list of the
+        same length as number of nodes.
+
+        :param param: parameters to check / prepare
+        :type param: list[native_type]|native_type|None
+        :param num_nodes: number of nodes in the network
+        :type num_nodes: int
+        """
+        if param is None:
+            param = [None] * num_nodes
+        if isinstance(param, native_type):
+            param = [param] * num_nodes
+        assert isinstance(param, list)
+        assert all(p is None or isinstance(p, native_type) for p in param)
+        assert len(param) == num_nodes
+        return param
+
+    def get_nested_parameters(self):
+        """
+        Return nested dictionary with parameters from all nodes and all masses
+        within this network.
+
+        :return: nested dictionary with all parameters
+        :rtype: dict
+        """
+        assert self.initialised
+        nested_dict = {NETWORK_NAME_STR: {}}
+        for node in self:
+            nested_dict[NETWORK_NAME_STR].update(node.get_nested_parameters())
+        nested_dict[NETWORK_CONNECTIVITY] = self.connectivity
+        nested_dict[NETWORK_DELAYS] = self.delays
+        return nested_dict
+
+    def init_network(self, **kwargs):
+        """
+        Initialise network and the nodes within.
+
+        :kwargs: optional keyword arguments to init_node
+        """
+        # create symbol for each node for input
+        self.sync_symbols = {
+            f"{symbol}_{node_idx}": se.Symbol(f"{symbol}_{node_idx}")
+            for symbol in self.sync_variables
+            for node_idx in range(self.num_nodes)
+        }
+        for node in self:
+            node.init_node(**kwargs)
+        assert all(node.initialised for node in self)
+        self.initialised = True
+
+    def update_parameters(self, parameters_dict):
+        """
+        Update parameters of this network, i.e. recursively for all nodes and
+        all masses.
+
+        :param parameters_dict: new parameters for the network
+        :type parameters_dict: dict
+        """
+        for node_key, node_params in parameters_dict.items():
+            if NODE_NAME_STR in node_key:
+                node_index = int(node_key.split("_")[-1])
+                assert node_index == self.nodes[node_index].index
+                self.nodes[node_index].update_parameters(node_params)
+            elif NETWORK_CONNECTIVITY == node_key:
+                assert node_params.shape == self.connectivity.shape
+                self.connectivity = node_params
+            elif NETWORK_DELAYS == node_key:
+                assert node_params.shape == self.delays.shape
+                self.delays = node_params
+            else:
+                logging.warning(f"Not sure what to do with {node_key}...")
+
+    def _callbacks(self):
```

Comment:
ignore `callbacks` and `numba_callbacks` for now - they are important for AdEx but not otherwise